numpy/core/src/multiarray/arrayobject.c

```diff
@@ -232,6 +232,9 @@ PyArray_SetBaseObject(PyArrayObject *arr, PyObject *obj)
     }
 
     ((PyArrayObject_fields *)arr)->base = obj;
+    if (PyArray_Check(obj)) {
+        PyArray_ENABLEFLAGS(obj, NPY_ARRAY_WARN_ON_WRITE);
```

Comment:
I had to set the same flag on `arr`, other than that this works pretty well. Thanks Matti.